libvast/src/arrow_table_slice.cpp

```diff
@@ -1277,4 +1278,155 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
+    },
+    [&](const std::shared_ptr<arrow::FixedSizeBinaryArray>& ba,
+        const std::shared_ptr<address_extension_type>& at)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(ba->byte_width() == 16);
+      return std::make_shared<address_array>(at, ba);
+    },
+    [&](const std::shared_ptr<arrow::FixedSizeBinaryArray>& ba,
+        const std::shared_ptr<subnet_extension_type>&)
+      -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(ba->byte_width() == 17);
+      auto subnet_array_data = convert_subnet_array(ba);
+      // TODO: the type is not actually correct, we have
+      // left:  struct<address: fixed_size_binary[16], length: uint8>
+      // right: struct<address: extension<vast.address>, length: uint8>
+      // so we can't create the subnet_array either (arrow asserts storage
+      // type) this gets "fixed" by arrow when re-reading the record batch
+      // in, but the record batch under construction exposes the wrong type
+      // information. to fix this, we require proper AddressBuilder /
+      // SubnetBuilder IIUC
+      // return std::make_shared<subnet_array>(st, subnet_array_data);
+      // this sorta works:  we need to disable verify_record_batch, though.
+      return subnet_array_data;
+    },
+    [&](const std::shared_ptr<arrow::UInt64Array>& a,
+        const std::shared_ptr<enum_extension_type>& et)
+      -> std::shared_ptr<arrow::Array> {
+      auto dict = make_dict_array(et->get_enum_type());
+      auto vals = arrow::compute::Cast(*a, arrow::int16());
+      auto dict_array = std::make_shared<arrow::DictionaryArray>(
+        arrow::dictionary(arrow::int16(), arrow::utf8()), vals.ValueUnsafe(),
+        dict);
+      return std::make_shared<enum_array>(et, dict_array);
+    },
+    [&](const std::shared_ptr<arrow::ListArray>& la,
+        const std::shared_ptr<arrow::MapType>& mt)
+      -> std::shared_ptr<arrow::Array> {
+      return std::make_shared<arrow::MapArray>(mt, la->length(),
+                                               la->value_offsets(),
+                                               la->values(), la->null_bitmap(),
+                                               la->null_count());
+    },
+    [&](const std::shared_ptr<arrow::Array>&,
+        const std::shared_ptr<arrow::StructType>& st)
+      -> std::shared_ptr<arrow::Array> {
+      // move back the iterator so it points to the first array in the struct
+      --array_iterator;
+      arrow::ArrayVector children{};
+      std::vector<std::string> field_names{};
+      children.reserve(st->num_fields());
+      field_names.reserve(st->num_fields());
+      for (const auto& field : st->fields()) {
+        children.push_back(make_arrow_array(array_iterator, field->type()));
+        field_names.push_back(field->name());
+      }
+      auto res = arrow::StructArray::Make(children, field_names);
+      return res.ValueOrDie();
+    },
+  };
+  return caf::visit(f, *array_iterator++, arrow_type);
```

Comment:
Does it maybe make more sense to only dispatch on the VAST type here, because we don't want to have to support the old Array's in the sum type access in the future anymore? We can then hardcode the static casts for the old Arrow encoding, because that is set in stone anyways.

Reply:
I'm not sure. I used the additional argument to the visitor only for some type safety and to avoid the explicit casting, and I have no strong opinion either way.

Reply:
changed according to the suggestion. It's annoying to explicitely cast, but then we can handle types that are not part of `sum_type_access` for array, in particular `FixedSizeBinaryArray`